Cache ConformanceAccessPaths in GSB::EquivalenceClass for big wins

jrose-apple opened the pull request:
This dramatically speeds up SubstitutionMap::lookupConformance, shaving off 1/3 of the time spent ~compiling the standard library~ generating the standard library module.

jrose-apple committed:
Cache ConformanceAccessPaths in GSB::EquivalenceClass for big wins

This dramatically speeds up SubstitutionMap::lookupConformance,
shaving off 1/3 of the time spent compiling the standard library.

jrose-apple commented:
@swift-ci Please test compiler performance

jrose-apple commented:
@swift-ci Please test

rjmccall commented on lib/AST/GenericSignature.cpp:
```diff
@@ -976,6 +981,7 @@ ConformanceAccessPath GenericSignature::getConformanceAccessPath(
     // Add the root of the path, which starts at this explicit requirement.
     path.path.push_back({rootType, conformingProto});
   };
+  buildPath = buildPathLambda;
```
I don't understand the purpose of binding `buildPath` specifically as a `function_ref` here.

jrose-apple replied:
That would probably work too. I didn't want to modify code I didn't understand too much, but the allocation was bothering me.

rjmccall commented on lib/AST/GenericSignature.cpp:
```diff
@@ -985,6 +991,7 @@ ConformanceAccessPath GenericSignature::getConformanceAccessPath(
   buildPath(getRequirements(), source, protocol, rootType, nullptr);
 
   // Return the path; we're done!
+  equivClass->conformanceAccessPathCache[protocol] = path;
```
If we're caching these instead of building them on-demand, can we reasonably return them by reference?  Or make `ConformanceAccessPath` just hold a reference to external storage so that it's cheaper to pass one around?

jrose-apple replied:
That seems sensible too.

jrose-apple commented:
So much so that I'm suspicious! How did it end up outputting less LLVM?

xedin commented:
One possible reason is - some of the projects failed while compiling with changes vs. without, is that right, @graydon?

jrose-apple commented:
It's supposed to ignore that, because if it didn't you'd get no useful data at all if one project fails. But that's my best guess too…

jrose-apple committed:
Move ConformanceAccessPaths to live completely on the ASTContext

John pointed out that if we're going to cache them, we might as well
go all the way and pass them around by reference.

jrose-apple committed:
Extract local helper to GenericSignature::buildConformanceAccessPath

It's a recursive helper that only captures one argument other than
'this'. Pay the cost of putting a declaration in a commonly-included
header to simplify the code.

No intended functionality change.

jrose-apple commented:
@swift-ci Please smoke test

rjmccall commented:
That looks great, thanks!

jrose-apple commented:
@swift-ci Please smoke test Linux

graydon commented:
@xedin @jrose-apple I don't know how this is happening (it is indeed supposed to be ignoring cases that failed when computing deltas!) but those stats are definitely indicative of a measurement error. If you look at the detailed-stats deltas, it's counting the intake of 43% less source-lines of code (AST.NumSourceLines); everything else follows from that.

graydon commented:
(Slightly more-nuanced read: the patch breaks non-WMO builds of a few projects. The actual improvement is seen in the release detailed counters -- Sema.NumGenericSignatureBuilders decreases by 7.5% -- with no improvement to time or space enough to register above threshold in the brief summary)

jrose-apple commented:
@swift-ci Please test source compatibility

jrose-apple commented:
Yep, looks like segfaults while building some things. Hm.

jrose-apple commented:
All right, let's try this again:

@swift-ci Please test compiler performance

graydon commented:
@swift-ci Please test compiler performance

jrose-apple commented:
There are still more projects failing to build in Debug mode. I'm looking into that today.

DougGregor reviewed (approved):
This looks great. I had this on a TODO list somewhere as a lower-priority optimization but never imagined it would be such a win. Thank you!

jrose-apple commented:
Okay, a few more of those Debug projects should pass now, but still not all of them, so I'm just going to merge this.
